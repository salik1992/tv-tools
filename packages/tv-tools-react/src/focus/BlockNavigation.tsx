import type { PropsWithChildren } from 'react';
import { FocusContext, useFocusContainer } from '../focus';

function markAsProcessed() {
	return true;
}

/**
 * BlockNavigation is a component that prevents navigation to escape from its children.
 * Useful for Modal, VirtualKeyboard, etc.
 */
export const BlockNavigation = ({ children }: PropsWithChildren) => {
	const { focusContextValue, useOnLeft, useOnUp, useOnDown, useOnRight } =
		useFocusContainer();

	useOnLeft(markAsProcessed);
	useOnRight(markAsProcessed);
	useOnUp(markAsProcessed);
	useOnDown(markAsProcessed);

	return (
		<FocusContext.Provider value={focusContextValue}>
			{children}
		</FocusContext.Provider>
	);
};
