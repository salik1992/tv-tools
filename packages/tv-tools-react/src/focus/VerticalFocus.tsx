import { type PropsWithChildren } from 'react';
import { useFocusContainer } from './useFocusContainer';

/**
 * Component for markup of where the focus should navigate vertically.
 * Direct focusable children will be navigated with up/down arrows in the order
 * they are placed in the DOM structure. This component does not render any
 * element.
 *
 * @prop id - optional id if you need to address it manually
 *
 * @example
 * ```typescriptreact
 * export const Menu = ({ ... }) => {
 *     // ...
 *     return (
 *          <div id="side-menu">
 *              <VerticalFocus>
 *                  <MenuItem to="home">Home</MenuItem>
 *                  <MenuItem to="search">Search</MenuItem>
 *                  <MenuItem to="epg">EPG</MenuItem>
 *                  <MenuItem to="profile">My Profile</MenuItem>
 *              </VerticalFocus>
 *          </div>
 *     )
 * }
 * ```
 * If the MenuItem are focusable (using Interactable component internally),
 * then when they are focused and we press up/down, the focus will move between
 * them in the order home <-> search <-> epg <-> profile.
 */
export const VerticalFocus = ({
	id,
	children,
}: PropsWithChildren<{ id?: string }>) => {
	const { FocusContextProvider, container, useOnUp, useOnDown } =
		useFocusContainer(id);

	useOnUp(
		(event) => {
			return container.moveFocus(-1, (event.target as HTMLElement).id);
		},
		[container],
	);

	useOnDown(
		(event) => {
			return container.moveFocus(1, (event.target as HTMLElement).id);
		},
		[container],
	);

	return <FocusContextProvider>{children}</FocusContextProvider>;
};
