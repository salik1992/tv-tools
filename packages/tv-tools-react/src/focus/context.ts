import { createContext } from 'react';
import { noop } from '@salik1992/tv-tools/utils/noop';
import type { Focus } from './types';

export const FocusContext = createContext<Focus>({
	addChild: noop,
});
