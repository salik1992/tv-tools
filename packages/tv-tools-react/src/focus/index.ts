export { FocusRoot } from './FocusRoot';
export { HorizontalFocus } from './HorizontalFocus';
export { Interactable } from './Interactable';
export { VerticalFocus } from './VerticalFocus';
export { useFocusContainer } from './useFocusContainer';
export { useKeySequenceListener } from './useKeySequenceListener';
