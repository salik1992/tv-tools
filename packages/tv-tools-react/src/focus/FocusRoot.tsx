import type { FocusEvent, KeyboardEvent, PropsWithChildren } from 'react';
import { useCallback } from 'react';
import { isDirectional } from '@salik1992/tv-tools/control';
import { focus } from '@salik1992/tv-tools/focus';

/**
 * The main root component for the focus management.
 * It creates a div element that is used for listening to key events in the manager.
 * Use at the root of your application.
 */
export const FocusRoot = ({
	children,
	/** Use this to prevent TV use their own navigation. */
	alwaysPreventNavigationalEvents,
}: PropsWithChildren<{
	alwaysPreventNavigationalEvents?: boolean;
}>) => {
	const onFocus = useCallback((event: FocusEvent) => {
		focus.handleFocusEvent(event.nativeEvent);
	}, []);

	const keyDownCapture = useCallback((event: KeyboardEvent<HTMLElement>) => {
		focus.handleKeyEvent('keydown', 'capture', event);
	}, []);

	const keyDownBubble = useCallback((event: KeyboardEvent<HTMLElement>) => {
		focus.handleKeyEvent('keydown', 'bubble', event);
		if (
			alwaysPreventNavigationalEvents &&
			isDirectional(event) &&
			document.activeElement?.tagName !== 'INPUT'
		) {
			event.preventDefault();
			event.stopPropagation();
		}
	}, []);

	const keyUpCapture = useCallback((event: KeyboardEvent<HTMLElement>) => {
		focus.handleKeyEvent('keyup', 'capture', event);
	}, []);

	const keyUpBubble = useCallback((event: KeyboardEvent<HTMLElement>) => {
		focus.handleKeyEvent('keyup', 'bubble', event);
		if (
			alwaysPreventNavigationalEvents &&
			isDirectional(event) &&
			document.activeElement?.tagName !== 'INPUT'
		) {
			event.preventDefault();
			event.stopPropagation();
		}
	}, []);

	return (
		<div
			id="focusRoot"
			onFocus={onFocus}
			onKeyDownCapture={keyDownCapture}
			onKeyDown={keyDownBubble}
			onKeyUpCapture={keyUpCapture}
			onKeyUp={keyUpBubble}
		>
			{children}
		</div>
	);
};
