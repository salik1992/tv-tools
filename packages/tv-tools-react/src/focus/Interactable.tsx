import {
	type ComponentProps,
	useEffect,
	useLayoutEffect,
	useMemo,
	useRef,
} from 'react';
import { Interactable as InteractableBase } from '@salik1992/tv-tools/focus';
import { useFocusContext } from './useFocusContext';

/**
 * The main component for rendering focusable elements. It renders a div and you
 * can pass all usual props that you could use for the div.
 *
 * @prop onPress - action that should happen when user presses ENTER or clicks on
 * the element. When given the Interactable works in uncontrolled mode.
 * @prop interactableBase - optional interactable base that will be used for this
 * Interactable. This is will make Interactable controlled.
 * @prop focusOnMount - optional mark to tell the component to focus itself when mounted
 * If controlled you can also force focus on update. Please note, if there are other
 * components using focusOnMount later in the document, they will take over.
 * @prop id - id that will be assigned to div and also used for focus management. If left
 * empty, the focus manager will assign auto id. Please note, the id parameter is required
 * for finding the origin of the event.
 * @prop tabIndex - if needed it can be overriden here, otherwise it is set to 0 by focus
 * manager
 *
 * @example
 * ```typescriptreact
 * const MovieTile = ({ data }) => {
 *     return (
 *         <Interactable className="movie-tile">
 *             <Image src={data.poster} />
 *             <H4>{data.title}</H4>
 *         </Interactable>
 *     )
 * }
 */
export const Interactable = ({
	onPress,
	interactableBase,
	focusOnMount,
	id,
	tabIndex,
	disabled = false,
	...props
}: (
	| {
			onPress?: undefined;
			interactableBase: InteractableBase;
	  }
	| {
			onPress: () => boolean;
			interactableBase?: undefined;
	  }
) & {
	focusOnMount?: boolean;
	disabled?: boolean;
} & ComponentProps<'div'>) => {
	const { focusManager, addChild } = useFocusContext();
	const interactable = useMemo(
		() =>
			disabled
				? null
				: (interactableBase ??
					new InteractableBase(focusManager, id, tabIndex)),
		[focusManager, id, disabled, interactableBase],
	);
	const ref = useRef<HTMLDivElement>(null);

	useEffect(() => {
		// Needs to run every time to maintain children order.
		if (interactable) {
			addChild(interactable.id);
		}
	});

	useEffect(() => {
		interactable?.setElement(ref.current);
	}, [interactable, ref.current]);

	useEffect(() => {
		if (focusOnMount && interactable) {
			focusManager.focus(interactable.id, { preventScroll: true });
		}
	}, [focusOnMount, interactable]);

	useLayoutEffect(
		() => () => {
			interactable?.destroy();
		},
		[interactable],
	);

	useEffect(() => {
		if (interactable && onPress) {
			interactable.setOnPress(onPress);
		}
	}, [interactable, onPress]);

	return <div {...props} ref={ref} />;
};
