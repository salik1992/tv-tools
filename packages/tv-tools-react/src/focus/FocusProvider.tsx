import type { ComponentProps, FocusEvent, KeyboardEvent } from 'react';
import { useCallback, useMemo } from 'react';
import { isDirectional } from '@salik1992/tv-tools/control';
import { FocusManager } from '@salik1992/tv-tools/focus';
import { noop } from '@salik1992/tv-tools/utils/noop';
import { FocusContext } from './context';

/**
 * The main root component for the focus management.
 * It creates a div element that is used for listening to key events in the manager and
 * it also creates the focus manager instance for the application which is available through
 * the FocusContext or useFocusManager hook.
 * Use at the root of your application.
 * @prop alwaysPreventNavigationalEvents - Use this to prevent TV use their own navigation.
 * @prop ...props - Any other props that you want to pass to the wrapping div element.
 */
export const FocusProvider = ({
	children,
	alwaysPreventNavigationalEvents,
	onFocus: onOuterFocus,
	onKeyDownCapture: onOuterKeyDownCapture,
	onKeyDown: onOuterKeyDown,
	onKeyUpCapture: onOuterKeyUpCapture,
	onKeyUp: onOuterKeyUp,
	...props
}: ComponentProps<'div'> & {
	/** Use this to prevent TV use their own navigation. */
	alwaysPreventNavigationalEvents?: boolean;
}) => {
	const focusManager = useMemo(() => new FocusManager(), []);

	const onFocus = useCallback(
		(event: FocusEvent<HTMLDivElement>) => {
			focusManager.handleFocusEvent(event.nativeEvent);
			onOuterFocus?.(event);
		},
		[focusManager, onOuterFocus],
	);

	const keyDownCapture = useCallback(
		(event: KeyboardEvent<HTMLDivElement>) => {
			focusManager.handleKeyEvent('keydown', 'capture', event);
			onOuterKeyDownCapture?.(event);
		},
		[focusManager, onOuterKeyDownCapture],
	);

	const keyDownBubble = useCallback(
		(event: KeyboardEvent<HTMLDivElement>) => {
			focusManager.handleKeyEvent('keydown', 'bubble', event);
			if (
				alwaysPreventNavigationalEvents &&
				isDirectional(event) &&
				document.activeElement?.tagName !== 'INPUT'
			) {
				event.preventDefault();
				event.stopPropagation();
			} else {
				onOuterKeyDown?.(event);
			}
		},
		[focusManager, alwaysPreventNavigationalEvents, onOuterKeyDown],
	);

	const keyUpCapture = useCallback(
		(event: KeyboardEvent<HTMLDivElement>) => {
			focusManager.handleKeyEvent('keyup', 'capture', event);
			onOuterKeyUpCapture?.(event);
		},
		[focusManager, onOuterKeyUpCapture],
	);

	const keyUpBubble = useCallback(
		(event: KeyboardEvent<HTMLDivElement>) => {
			focusManager.handleKeyEvent('keyup', 'bubble', event);
			if (
				alwaysPreventNavigationalEvents &&
				isDirectional(event) &&
				document.activeElement?.tagName !== 'INPUT'
			) {
				event.preventDefault();
				event.stopPropagation();
			} else {
				onOuterKeyUp?.(event);
			}
		},
		[focusManager, alwaysPreventNavigationalEvents, onOuterKeyUp],
	);

	const focusContextValue = useMemo(
		() => ({
			focusManager,
			addChild: noop,
		}),
		[focusManager],
	);

	return (
		<div
			{...props}
			onFocus={onFocus}
			onKeyDownCapture={keyDownCapture}
			onKeyDown={keyDownBubble}
			onKeyUpCapture={keyUpCapture}
			onKeyUp={keyUpBubble}
		>
			<FocusContext.Provider value={focusContextValue}>
				{children}
			</FocusContext.Provider>
		</div>
	);
};
