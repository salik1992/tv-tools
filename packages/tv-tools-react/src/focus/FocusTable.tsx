import {
	type ComponentProps,
	type ComponentType,
	type PropsWithChildren,
	type ReactNode,
	createContext,
	useCallback,
	useContext,
	useEffect,
	useMemo,
} from 'react';
import type { TableFocusContainer } from '@salik1992/tv-tools/focus';
import { FocusContext } from './context';
import { useTableFocusContainer } from './useTableFocusContainer';

const ChildrenOnlyComponent = ({ children }: PropsWithChildren) => {
	return <>{children}</>;
};

const TableContext = createContext<{
	container?: TableFocusContainer;
	TrComponent: ComponentType<PropsWithChildren<unknown>>;
	TdComponent: ComponentType<PropsWithChildren<unknown>>;
}>({
	TrComponent: ChildrenOnlyComponent,
	TdComponent: ChildrenOnlyComponent,
});

const FocusTd = ({
	rowSpan,
	colSpan,
	children,
	...props
}: PropsWithChildren<{
	colSpan?: number;
	rowSpan?: number;
}>) => {
	const { addChild, ...originalfocusContextValue } = useContext(FocusContext);
	const { TdComponent } = useContext(TableContext);

	const addChildWithSpans = useCallback(
		(childId: string) => {
			addChild(childId, { rowSpan, colSpan });
		},
		[addChild, rowSpan, colSpan],
	);

	const focusContextValue = useMemo(
		() => ({ ...originalfocusContextValue, addChild: addChildWithSpans }),
		[addChildWithSpans, originalfocusContextValue],
	);

	return (
		<FocusContext.Provider value={focusContextValue}>
			<TdComponent {...props}>{children}</TdComponent>
		</FocusContext.Provider>
	);
};

const FocusTr = (props: PropsWithChildren<unknown>) => {
	const { TrComponent, container } = useContext(TableContext);

	useEffect(() => {
		// Needs to run every time to maintain children order.
		container?.addRow();
	});

	return <TrComponent {...props} />;
};

export type FocusTableRenderComponents<
	TrComponent extends ComponentType,
	TdComponent extends ComponentType,
> = {
	FocusTr: (props: ComponentProps<TrComponent>) => ReactNode;
	FocusTd: (
		props: {
			colSpan?: number;
			rowSpan?: number;
		} & ComponentProps<TdComponent>,
	) => ReactNode;
};

export const FocusTable = ({
	id,
	children,
	// @ts-expect-error: TODO
	TableComponent = ChildrenOnlyComponent,
	TrComponent = ChildrenOnlyComponent,
	TdComponent = ChildrenOnlyComponent,
	...tableProps
}: {
	children: (
		components: FocusTableRenderComponents<
			typeof TrComponent,
			typeof TdComponent
		>,
		container: TableFocusContainer,
	) => ReactNode;
	id?: string;
	TableComponent?: ComponentType;
	TrComponent?: ComponentType;
	TdComponent?: ComponentType;
} & ComponentProps<typeof TableComponent>) => {
	const {
		focusContextValue,
		container,
		useOnLeft,
		useOnUp,
		useOnDown,
		useOnRight,
	} = useTableFocusContainer(id);

	useOnLeft(
		(event) =>
			container.moveFocus({ x: -1 }, (event.target as HTMLElement).id),
		[container],
	);

	useOnRight(
		(event) =>
			container.moveFocus({ x: 1 }, (event.target as HTMLElement).id),
		[container],
	);

	useOnUp(
		(event) =>
			container.moveFocus({ y: -1 }, (event.target as HTMLElement).id),
		[container],
	);

	useOnDown(
		(event) =>
			container.moveFocus({ y: 1 }, (event.target as HTMLElement).id),
		[container],
	);

	const tableContextValue = useMemo(
		() => ({
			container,
			TrComponent,
			TdComponent,
		}),
		[TrComponent, TdComponent],
	);

	return (
		<TableContext.Provider value={tableContextValue}>
			<FocusContext.Provider value={focusContextValue}>
				<TableComponent {...tableProps}>
					{children({ FocusTr, FocusTd }, container)}
				</TableComponent>
			</FocusContext.Provider>
		</TableContext.Provider>
	);
};
