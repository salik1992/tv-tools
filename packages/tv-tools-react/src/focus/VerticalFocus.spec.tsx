import { render, act } from '@testing-library/react';
import { UP, DOWN } from '@salik1992/tv-tools/control';
import { focus } from '@salik1992/tv-tools/focus';
import { FocusRoot } from './FocusRoot';
import { Interactable } from './Interactable';
import { VerticalFocus } from './VerticalFocus';

const onPress = jest.fn();
const focusSpy = jest.spyOn(focus, 'focus');

describe('HorizontalFocus', () => {
	beforeEach(() => {
		focusSpy.mockClear();
	});

	it('should render the children', () => {
		const { container, unmount } = render(
			<FocusRoot>
				<VerticalFocus id="hf">
					<Interactable id="i1" onPress={onPress} />
					<Interactable id="i2" onPress={onPress} />
					<Interactable id="i3" onPress={onPress} />
					<Interactable id="i4" onPress={onPress} />
				</VerticalFocus>
			</FocusRoot>,
		);
		expect(container.innerHTML).toMatchSnapshot();
		unmount();
	});

	it('should move focus with UP/DOWN keys', () => {
		const { container, unmount } = render(
			<FocusRoot>
				<VerticalFocus id="hf">
					<Interactable id="i1" onPress={onPress} focusOnMount />
					<Interactable id="i2" onPress={onPress} />
					<Interactable id="i3" onPress={onPress} />
				</VerticalFocus>
			</FocusRoot>,
		);
		expect(focusSpy).toHaveBeenNthCalledWith(1, 'i1', {
			preventScroll: true,
		});

		// DOWN
		let event = DOWN.toKeyboardEvent('keydown');
		let spyStopPropagation = jest.spyOn(event, 'stopPropagation');
		let spyPreventDefault = jest.spyOn(event, 'preventDefault');
		act(() => {
			container.querySelector('#i1')?.dispatchEvent(event);
		});
		expect(focusSpy).toHaveBeenNthCalledWith(2, 'i2', {
			preventScroll: true,
		});
		expect(spyStopPropagation).toHaveBeenCalled();
		expect(spyPreventDefault).toHaveBeenCalled();

		// DOWN
		event = DOWN.toKeyboardEvent('keydown');
		spyStopPropagation = jest.spyOn(event, 'stopPropagation');
		spyPreventDefault = jest.spyOn(event, 'preventDefault');
		act(() => {
			container.querySelector('#i2')?.dispatchEvent(event);
		});
		expect(focusSpy).toHaveBeenNthCalledWith(3, 'i3', {
			preventScroll: true,
		});
		expect(spyStopPropagation).toHaveBeenCalled();
		expect(spyPreventDefault).toHaveBeenCalled();

		// DOWN
		event = DOWN.toKeyboardEvent('keydown');
		spyStopPropagation = jest.spyOn(event, 'stopPropagation');
		spyPreventDefault = jest.spyOn(event, 'preventDefault');
		act(() => {
			container.querySelector('#i3')?.dispatchEvent(event);
		});
		expect(focusSpy).toHaveBeenCalledTimes(3);
		expect(spyStopPropagation).not.toHaveBeenCalled();
		expect(spyPreventDefault).not.toHaveBeenCalled();

		// UP
		event = UP.toKeyboardEvent('keydown');
		spyStopPropagation = jest.spyOn(event, 'stopPropagation');
		spyPreventDefault = jest.spyOn(event, 'preventDefault');
		act(() => {
			container.querySelector('#i3')?.dispatchEvent(event);
		});
		expect(focusSpy).toHaveBeenNthCalledWith(4, 'i2', {
			preventScroll: true,
		});
		expect(spyStopPropagation).toHaveBeenCalled();
		expect(spyPreventDefault).toHaveBeenCalled();

		// UP
		event = UP.toKeyboardEvent('keydown');
		spyStopPropagation = jest.spyOn(event, 'stopPropagation');
		spyPreventDefault = jest.spyOn(event, 'preventDefault');
		act(() => {
			container.querySelector('#i2')?.dispatchEvent(event);
		});
		expect(focusSpy).toHaveBeenNthCalledWith(5, 'i1', {
			preventScroll: true,
		});
		expect(spyStopPropagation).toHaveBeenCalled();
		expect(spyPreventDefault).toHaveBeenCalled();

		// UP
		event = UP.toKeyboardEvent('keydown');
		spyStopPropagation = jest.spyOn(event, 'stopPropagation');
		spyPreventDefault = jest.spyOn(event, 'preventDefault');
		act(() => {
			container.querySelector('#i1')?.dispatchEvent(event);
		});
		expect(focusSpy).toHaveBeenCalledTimes(5);
		expect(spyStopPropagation).not.toHaveBeenCalled();
		expect(spyPreventDefault).not.toHaveBeenCalled();

		unmount();
	});
});
