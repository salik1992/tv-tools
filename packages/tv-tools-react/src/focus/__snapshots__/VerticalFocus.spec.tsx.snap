// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HorizontalFocus should render the children 1`] = `"<div id="focusRoot"><div id="i1" tabindex="0"></div><div id="i2" tabindex="0"></div><div id="i3" tabindex="0"></div><div id="i4" tabindex="0"></div></div>"`;
